stdlib/public/runtime/SwiftDtoa.cpp

```diff
@@ -624,9 +624,184 @@ int swift_decompose_double(double d,
 }
 #endif
 
+#if SWIFT_DTOA_FLOAT16_SUPPORT
+// Decompose an IEEE 754 binary16 half-precision float
+// into decimal digits and a corresponding decimal exponent.
+
+// See swift_decompose_double for detailed comments on the algorithm here
+// This can almost certainly be simplified a great deal.  This
+// first iteration just copies the code from float.
+int swift_decompose_float16(const __fp16 *f,
+    int8_t *digits, size_t digits_length, int *decimalExponent)
+{
+    static const int significandBitCount = 10;
+    static const uint32_t significandMask
+        = ((uint32_t)1 << significandBitCount) - 1;
+    static const int exponentBitCount = 5;
+    static const int exponentMask = (1 << exponentBitCount) - 1;
+    // See comments in swift_decompose_double
+    static const int64_t exponentBias = (1 << (exponentBitCount - 1)) - 2; // 14
+
+    // Step 0: Deconstruct the target number
+    // Note: this strongly assumes IEEE 754 binary16 format
```

Comment:
__fp16 is a clang extension and is always binary16, so this is not an assumption.

Reply:
Clarified the comment.